### PR TITLE
Cordova: support for local plugins

### PR DESCRIPTION
**Pull request by wojtkowiak**
I'm working on project which heavily depends on a cordova plugin which is also under development. Without support for plugins added from local path it was really hard to work on simultaneously on both meteor project and cordova plugin.

Supported syntax is:
meteor add cordova:com.my.plugin@file://../../my/relative/path
meteor add cordova:com.my.plugin@file:///home/user/my/absolute/path

Because I did not want to check whether the plugin has changed I am refreshing the plugin on each build/run. This done by issuing remove and adding the plugin again.
Because cordova 4.2.0 is not properly handling removing plugins added from local path I had to switch to 4.3.0 in which this issue is resolved.

I see that some discussion have been going on here https://github.com/meteor/meteor/issues/2925
Let me know if there is any chance this could be accepted and if yes I am looking forward for comments on this.

### CODE REVIEW

File: tools/commands-cordova.js [hunk 1]
```diff
@@ -464,8 +471,16 @@ var targetsToPlatforms = cordova.targetsToPlatforms = function (targets) {
 var installPlugin = function (cordovaPath, name, version, conf) {
   verboseLog('Installing a plugin', name, version);
 
+  var pluginInstallCommand;
+
+  if (utils.isUrlWithFileUri(version)) {
+    // Strip file://
+    pluginInstallCommand = version.substr(7);
```
Comment (Slava) [inline tools/commands-cordova.js]: replace `7` with `"file://".length` or something more readable

File: tools/commands-cordova.js [hunk 2]
```diff
@@ -464,8 +471,16 @@ var targetsToPlatforms = cordova.targetsToPlatforms = function (targets) {
 var installPlugin = function (cordovaPath, name, version, conf) {
   verboseLog('Installing a plugin', name, version);
 
+  var pluginInstallCommand;
+
+  if (utils.isUrlWithFileUri(version)) {
```
Comment (Slava) [inline tools/commands-cordova.js]: rename this utility function to `isUrlWithFileScheme`

File: tools/commands-cordova.js [hunk 3]
```diff
@@ -614,17 +629,30 @@ var ensureCordovaPlugins = function (projectContext, options) {
...
   // upgrade the version on an individual Cordova plugin. Instead, whenever a
   // new Cordova plugin is added or removed, or its version is changed,
-  // we just reinstall all of the plugins.
+  // we just reinstall all of the plugins. Additionally if there are any plugins
+  // added from the local path, we will reinstall them just to be sure they
+  // are up to date since we do not track changes in their sources.
 
-  var shouldReinstallPlugins = false;
+  var shouldReinstallPlugins = false,
+      shouldReinstallPluginsFromLocalPaths = false,
```
Comment (Slava) [inline tools/commands-cordova.js]: why is it a separate variable? do we even need this distinction here?

File: scripts/dev-bundle-tool-package.js [hunk 4]
```diff
@@ -52,7 +52,7 @@ var packageJson = {
     // 2.4.0 (more or less, the package.json change isn't committed) plus our PR
     // https://github.com/williamwicks/node-eachline/pull/4
     eachline: "https://github.com/meteor/node-eachline/tarball/ff89722ff94e6b6a08652bf5f44c8fffea8a21da",
-    cordova: "4.2.0",
+    cordova: "4.3.0",
```
Comment (Slava) [inline scripts/dev-bundle-tool-package.js]: This should be a separate commit.

File: tools/tests/apps/package-tests/packages/empty-cordova-plugin/package.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Package.describe({
+  version: "1.0.0",
+  summary: "contains a empty cordova plugin"
+});
+
+Package.onUse(function(api) {
+  Cordova.depends({
+    'com.cordova.empty': 'file://../../../../cordova-local-plugin'
```
Comment (Slava) [inline tools/tests/apps/package-tests/packages/empty-cordova-plugin/package.js]: can you put a comment? What is this relative path leading to?

**Comment by Slava** [comment]
@wojtkowiak This is a great PR! Thank you for this.

I had some comments inline, I like that there are tests :+1:

File: tools/commands-cordova.js (hunk 1 above)
Comment (wojtkowiak) [inline tools/commands-cordova.js]: Fixed with `"file://".length`

File: tools/commands-cordova.js (hunk 2 above)
Comment (wojtkowiak) [inline tools/commands-cordova.js]: Done

**Comment by wojtkowiak** [comment]
Firstly, thank for you comments 😏 

About the `shouldReinstallPluginsFromLocalPaths` var. 
The whole concept is that if you are using a plugin from a local path, then probably it is because you are also developing the plugin in the meantime. So if we are not checking if the sources of the plugin have changed, we need to reinstall it on every build just to be sure it is up to date. 
For this I needed to see if we have any plugins added from the local path just to know if we need to reinstall those plugins on this run/build. However this var is unnecessary - I have deleted it and instead I am just checking if there are any local plugins added using `if (pluginsFromLocalPath.length > 0)`.

I guess it would be also nice to have an option to disable this behavior (reinstallation of local cordova plugins on each run/build) if for some reason this is not desirable. This could be done for example by adding some flag like `meteor add cordova:com.plugin@file://path/dir --disable-auto-reinstall`. If you want I can prepare that too.

About the relative path `meteor add cordova:com.my.plugin@file://../../my/relative/path` - this needs to be a relative path from meteor_application/.meteor/local/cordova-build because that's where the cordova command is executed.

**Comment by wojtkowiak** [comment]
Should I also take care of history.md and https://github.com/meteor/meteor/wiki/Meteor-Cordova-Phonegap-integration?

**Comment by Slava** [comment]
> About the relative path meteor add cordova:com.my.plugin@file://../../my/relative/path - this needs to be a relative path from meteor_application/.meteor/local/cordova-build because that's where the cordova command is executed.

I don't think this is good. The relative url should not be based off the internal build folder, rather the user should pass a relative url to the Meteor app folder. Later you can re-calculate it for the cordova plugin, internally.

**Comment by wojtkowiak** [comment]
I agree, having it relative to Meteor app seems like a much better idea - changed.
I am thinking maybe I should add some kind of sanity check for situations like adding path like that `@file://home/user/....` which will be resolved as relative. I guess actually nobody would use it like `@file://plugin_dir` because having the plugin in the project dir would make meteor try to build at least the js files. So maybe we should permit only two possibilities: `@file:///` and `@file://../` and treat anything else as invalid. Let me know what do you think about this and about this `--disable-auto-reinstall` idea.

**Comment by qdsang** [comment]
This function is great.
In China Cordova installation service is not very stable.

**Comment by wojtkowiak** [comment]
I have done some additional testing on plain Xubuntu 14.04 and OS X 10.10 and one thing came up. Cordova 4.3.0 seems to force android-21 target. I see two quick fixes for that:
1. after cordova adding android platform we can automatically change the Android manifest and project.properties files to change the target to android-21 to android-19
2. download only (or also) target-21 when doing `install-sdk android`

However maybe this should be addressed in more general manner. I would propose to add `--target-sdk` (or `--target-sdk-level`) to `install-sdk android` and the same option to `add-platform android`. And maybe this could also be configurable in the mobile-config.js. 
It is a pitty that cordova currently does not provide a nice way of setting this, we would have to edit the files anyway just to set the sdk level - or maybe I am not aware of any nice way to set it.

**Comment by wojtkowiak** [comment]
I have done some additional testing on plain Xubuntu 14.04 and OS X 10.10 and one thing came up. Cordova 4.3.0 is now using android-21 target and it does not compile on android-19 so I changed it to 21.

Maybe it would be a good idea to allow providing a cordova-android version to `meteor add-platform android`. Plain corodova allows it. This would be usefull as you would be able to force cordova-android 4.0.0. in which crosswalk is added as a plugin and now works out of the box :)

**Comment by Exilz** [comment]
It would be fantastic to be able to use cordova-android 4.0, yes.

File: tools/commands-cordova.js [hunk 6]
```diff
@@ -638,38 +673,64 @@ var ensureCordovaPlugins = function (projectContext, options) {
...
-  if (shouldReinstallPlugins) {
+  if (! _.isEmpty(pluginsFromLocalPath))
+    verboseLog('Reinstalling cordova plugins added from the local path');
+
+  if (shouldReinstallPlugins || ! _.isEmpty(pluginsFromLocalPath)) {
     // Loop through all of the current plugins and remove them one by one until
-    // we have no plugins. It's necessary to loop because we might have
-    // dependencies between plugins.
-    var uninstallAllPlugins = function () {
+    // we have deleted proper amount of plugins. It's necessary to loop because
```
Comment (Slava) [inline tools/commands-cordova.js]: Can you explain this? What is "proper amount of plugins"? Why is it better or more correct than "delete all plugins"?

**Comment by Slava** [comment]
Merged to the branch called `cordova5-update`. So far so good. Thank you for the PR.

**Comment by Slava** [comment]
And yeah, I am still waiting for the replies on a couple of comments inline.

File: tools/commands-cordova.js (hunk 6 above)
Comment (wojtkowiak) [inline tools/commands-cordova.js]: Basically the `uninstallPlugins` function now gets the `pluginsToUninstall` array. Now it can delete any amount of plugins so it is not allways all of them. By "proper" I meant that the amount of deleted plugins is equal to size of the array :) Anyway we could just say 'until we have deleted all of the plugins provided in pluginsToUninstall array' or sth like that. The two scenarios (delete all or just the plugins from the local path) are described in the comment on 639, maybe it should also be reminded before the conditional on 711.

Comment (Slava) [inline tools/commands-cordova.js]: Any reason why we would uninstall only some plugins and not all of them? Previously we noticed that failures on uninstalling a couple of plugins and installing the rest are pretty often and a "safe" way would be uninstalling everything at once.

**Comment by wojtkowiak** [comment]
Comments answered. I see that you've already merged it into devel :+1: Glad that I could contribute :)

File: tools/commands-cordova.js (hunk 6 above)
Comment (wojtkowiak) [inline tools/commands-cordova.js]: Ok, maybe I have not explained it clearly. Before this PR like you said if there was any change (added/remove plugins) we've been deleting all the plugins because it was "safe". This behavior was not changed. Still if you will add/remove any plugins to the project it will delete all of them and install them again just like it did before. 
Since we need to refresh the local plugins (as we do not track changes in their sources) on every build I wanted to reinstall only the local plugins and that is why I needed to modify this function to support this case. On line 711 you can see that I am either reinstalling all of the plugins or just those added from local path so I am never selectively reinstalling a plugin that is added from url or cordova repo - so the "safe" way of ensuring the good state of cordova build was retained.

Comment (Slava) [inline tools/commands-cordova.js]: got it

**Comment by devinesoftware** [comment]
I am trying this with the latest version of meteor and I am getting the following message:

"Must declare exact version of dependency"

What am I doing wrong?

**Comment by codeluggage** [comment]
I am also having this problem, running meteor 1.1.0.3. 

I've tried to change the syntax for how it is written, adding in the version number before, between and after, but it all results in errors or this message: 

"~/path/to/project/.meteor/cordova-plugins: Cordova plugin must specify version:" 

Is there anything we can do to debug, test or check this issue so we can find a workaround or solution @wojtkowiak ? 

Regardless, thank you so much for this work! Using a tarball from github is not the smoothest workflow, hehe. I'm looking forward to this working again!

**Comment by wojtkowiak** [comment]
@devinesoftware @codeluggage supported syntax is 
`meteor add cordova:com.my.plugin@file://../../my/relative/path`
`meteor add cordova:com.my.plugin@file:///home/user/my/absolute/path`

You need to use the latest devel version of meteor or the current 1.2 RC.
If you want to use devel then you neeed to checkout devel branch and run the `meteor` command from the checkout. So if you are in you project dir instead of typing `meteor` you type the relative or absolute path to the checkout like: `/home/omega/projects/meteor/meteor run`
To use RC you need to update your project using `meteor update --release METEOR@1.2-rc.15` Pay attention as new rc version is published very often.

**Comment by obiwarn** [comment]
Thank you.

Is this documented anywhere?

**Comment by wojtkowiak** [comment]
Hmm seems that it should be mentioned here: https://github.com/meteor/meteor/wiki/Meteor-Cordova-integration but it is not. Maybe it would be good to file an issue for that.

**Comment by markoshust** [comment]
I agree that this should be documented at https://github.com/meteor/meteor/wiki/Meteor-Cordova-integration -- not familiar how wiki pages work, can we make a pull request there?

**Comment by markoshust** [comment]
cc @stubailo
